crates/runtime/src/libcalls.rs

```diff
@@ -7,99 +7,253 @@
 //!
-//!       // Now that we did `clone_from_raw`, it is safe to do a GC (or do
-//!       // anything else that might transitively GC, like call back into
-//!       // Wasm!)
-//!   }
-//!   ```
+//! Avoid this by eagerly converting into a proper `VMExternRef`! (Unfortunately
+//! there is no macro to help us automatically get this correct, so stay
+//! vigilant!)
+//!
+//! ```ignore
+//! pub unsafe extern "C" my_libcall_takes_ref(raw_extern_ref: *mut u8) {
+//!     // Before `clone_from_raw`, `raw_extern_ref` is potentially unrooted,
+//!     // and doing GC here could lead to use after free!
+//!
+//!     let my_extern_ref = if raw_extern_ref.is_null() {
+//!         None
+//!     } else {
+//!         Some(VMExternRef::clone_from_raw(raw_extern_ref))
+//!     };
+//!
+//!     // Now that we did `clone_from_raw`, it is safe to do a GC (or do
+//!     // anything else that might transitively GC, like call back into
+//!     // Wasm!)
+//! }
+//! ```
 
 use crate::externref::VMExternRef;
 use crate::instance::Instance;
 use crate::table::{Table, TableElementType};
-use crate::traphandlers::{raise_lib_trap, raise_user_trap, resume_panic};
+use crate::traphandlers::resume_panic;
 use crate::vmcontext::{VMCallerCheckedAnyfunc, VMContext};
 use std::mem;
 use std::ptr::{self, NonNull};
 use wasmtime_environ::{
     DataIndex, ElemIndex, FuncIndex, GlobalIndex, MemoryIndex, TableIndex, TrapCode,
 };
 
-/// Implementation of memory.grow for locally-defined 32-bit memories.
-pub unsafe extern "C" fn memory32_grow(
-    vmctx: *mut VMContext,
-    delta: u64,
-    memory_index: u32,
-) -> *mut u8 {
-    // Memory grow can invoke user code provided in a ResourceLimiter{,Async},
-    // so we need to catch a possible panic
-    let ret = match std::panic::catch_unwind(|| {
+/// Define a libcall function.
+///
+/// This macro takes care of handling panics and raising traps for you, without
+/// you having to worry about leaking resources due to `longjmp`ing and skipping
+/// destructors. Additionally, it handles defining the Wasm-to-host trampoline
+/// for maintaining Wasm frame and stack pointers, used when capturing
+/// backtraces.
+///
+/// What kind of return should my libcall have if it raises traps?
+///
+/// * If your libcall can raise a user trap, use `-> Result<T>` for returns
+///   (i.e. the error variant is the default `anyhow::Error).
+///
+/// * However, if it can only raise a lib trap, then use `-> Result<T,
+///   TrapCode>` for your return type (i.e. use the `wasmtime_environ::TrapCode`
+///   type for the error variant).
+///
+/// * If your libcall can both raise user and lib traps, then use `-> Result<T,
+///   TrapReason>`.
+///
+/// * If your libcall is infallible, just use `-> T` directly, as usual.
+///
+/// ### Example
+///
+/// ```ignore
+/// libcall! {
+///     my_libcall => unsafe fn impl_my_libcall(vmctx: *mut VMContext) -> Result<u64> {
+///         // do stuff here
+///     }
+/// }
+/// ```
+macro_rules! libcall {
```

Comment:
Personally I feel that this macro is still too complicated for its own good, but I sent you https://github.com/fitzgen/wasmtime/pull/7 which is what I was thinking of how to simplify it.